key_test.go

```diff
@@ -191,34 +384,236 @@ func TestReadInput(t *testing.T) {
+		{"ctrl+@",
+			[]byte{'\x00'},
+			[]Msg{KeyMsg{Type: KeyCtrlAt}},
+		},
+		{"alt+ctrl+@",
+			[]byte{'\x1b', '\x00'},
+			[]Msg{KeyMsg{Type: KeyCtrlAt, Alt: true}},
+		},
+		{"esc",
+			[]byte{'\x1b'},
+			[]Msg{KeyMsg{Type: KeyEsc}},
+		},
+		{"alt+esc",
+			[]byte{'\x1b', '\x1b'},
+			[]Msg{KeyMsg{Type: KeyEsc, Alt: true}},
+		},
+		// Bracketed paste does not work yet.
+		{"?CSI[50 48 48 126]? a   b ?CSI[50 48 49 126]?",
+			[]byte{
+				'\x1b', '[', '2', '0', '0', '~',
+				'a', ' ', 'b',
+				'\x1b', '[', '2', '0', '1', '~'},
+			[]Msg{
+				// What we expect once bracketed paste is recognized properly:
+				//
+				//  KeyMsg{Type: KeyRunes, Runes: []rune("a b")},
+				//
+				// What we get instead (for now):
+				unknownCSISequenceMsg{0x1b, 0x5b, 0x32, 0x30, 0x30, 0x7e},
+				KeyMsg{Type: KeyRunes, Runes: []rune{'a'}},
+				KeyMsg{Type: KeySpace, Runes: []rune{' '}},
+				KeyMsg{Type: KeyRunes, Runes: []rune{'b'}},
+				unknownCSISequenceMsg{0x1b, 0x5b, 0x32, 0x30, 0x31, 0x7e},
+			},
+		},
+	}
+	if runtime.GOOS != "windows" {
+		// Sadly, utf8.DecodeRune([]byte(0xfe)) returns a valid rune on windows.
+		// This is incorrect, but it makes our test fail if we try it out.
+		testData = append(testData,
+			test{"?0xfe?",
+				[]byte{'\xfe'},
+				[]Msg{unknownInputByteMsg(0xfe)},
+			},
+			test{"a ?0xfe?   b",
+				[]byte{'a', '\xfe', ' ', 'b'},
+				[]Msg{
+					KeyMsg{Type: KeyRunes, Runes: []rune{'a'}},
+					unknownInputByteMsg(0xfe),
+					KeyMsg{Type: KeySpace, Runes: []rune{' '}},
+					KeyMsg{Type: KeyRunes, Runes: []rune{'b'}},
+				},
+			},
+		)
+	}
+
+	for i, td := range testData {
 		t.Run(fmt.Sprintf("%d: %s", i, td.keyname), func(t *testing.T) {
 			msgs, err := readInputs(bytes.NewReader(td.in))
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}
+
+			// Compute the title for the event sequence.
+			var buf strings.Builder
+			for i, msg := range msgs {
+				if i > 0 {
+					buf.WriteByte(' ')
+				}
+				if s, ok := msg.(fmt.Stringer); ok {
+					buf.WriteString(s.String())
+				} else {
+					fmt.Fprintf(&buf, "%#v:%T", msg, msg)
+				}
+			}
+			title := buf.String()
+			if title != td.keyname {
+				t.Errorf("expected message titles:\n  %s\ngot:\n  %s", td.keyname, title)
+			}
+
 			if len(msgs) != len(td.out) {
-				t.Fatalf("unexpected message list length")
+				t.Fatalf("unexpected message list length: got %d, expected %d\n%#v", len(msgs), len(td.out), msgs)
 			}
 
-			if len(msgs) == 1 {
-				if m, ok := msgs[0].(KeyMsg); ok && m.String() != td.keyname {
-					t.Fatalf(`expected a keymsg %q, got %q`, td.keyname, m)
-				}
+			if !reflect.DeepEqual(td.out, msgs) {
+				t.Fatalf("expected:\n%#v\ngot:\n%#v", td.out, msgs)
 			}
+		})
+	}
+}
 
-			for i, v := range msgs {
-				if m, ok := v.(KeyMsg); ok &&
-					m.String() != td.out[i].(KeyMsg).String() {
-					t.Fatalf(`expected a keymsg %q, got %q`, td.out[i].(KeyMsg), m)
+// randTest defines the test input and expected output for a sequence
```

Comment:
Good enough for now, but I guess we could eventually replace the randomizer with a proper fuzzer?

Reply:
Yes that is reasonable. Do you want to do this right away? Is there something standard we can use?

Reply:
Let's keep it as it is, the code looks and works fine for me. As soon as we bump Go to 1.18 we can start moving to the Go fuzzer that's part of the standard toolchain: https://go.dev/security/fuzz/